### PR TITLE
Fix #4956: autodoc: Failed to extract document from a subclass of the class on mocked module

### LINKED ISSUE #4956: Classes inheriting from other classes, defined in mocked modules, are ignored
As advised by @djhoese, this is a reopening of #1880.
Unfortunately, this is not #2174 (solved).
Actually, the docstrings of classes inheriting from other classes in mocked modules are ignored.
A real case is described in https://stackoverflow.com/q/50203452/3565696.
Here is a minimal working example, all files in the same directory
(also attached as [sphinx-mock-bug.zip](https://github.com/sphinx-doc/sphinx/files/1993341/sphinx-mock-bug.zip), for convenience).

Project module `foo.y`:
```
import bar

class FooClass(bar.BarClass):
    """This is myclass"""
    pass
```
`conf.py`:
```
import sys
import os

# The FooClass docstring shows up only if this line is commented out
autodoc_mock_imports = ['bar']

sys.path.insert(0, os.path.abspath('.'))

extensions = [
    'sphinx.ext.autodoc',
]
```
`contents.rst`:
```
foo
===

.. automodule:: foo
    :members:
    :noindex:
```
And finally, the`bar.py` "external" module (only used when the autodoc_mock_imports is commented out):
```
class BarClass():
	pass
```

The command used to build:
`sphinx-build  -b html . _build/html`

To see the file where the dosctring should be visible:
`firefox _build/html/foo.html`

Environment info:
- Linux (openSUSE 42.2)
- Python version: 3.4.6 and 3.6.4
- Sphinx version: 1.7.4
[and same on readthedocs for the real project]

### PR DESCRIPTION
**Pull request by tk0miya**
### Feature or Bugfix
- Bugfix

### Purpose
- see #4956

### CODE REVIEW

File: sphinx/ext/autodoc/importer.py [hunk 1]
```diff
@@ -31,6 +31,14 @@
 class _MockObject(object):
     """Used by autodoc_mock_imports."""
 
+    def __new__(cls, *args, **kwargs):
+        # type: (Any, Any) -> Any
+        if len(args) == 3 and isinstance(args[1], tuple) and args[1][-1].__class__ is cls:
+            # subclassing MockObject
+            return type(args[0], (_MockObject,), args[2], **kwargs)  # type: ignore
```
Comment (shimizukawa) [inline sphinx/ext/autodoc/importer.py:38]: confirmed.

**Review by shimizukawa — APPROVED** [review]
LGTM!

File: sphinx/ext/autodoc/importer.py [hunk 2]
```diff
@@ -31,6 +31,14 @@
...
+        else:
+            return super().__new__(cls)  # type: ignore
```
Comment (shimizukawa) [inline sphinx/ext/autodoc/importer.py]: not works with py2. Is `super(M, cls)` correct?
```
            return super(M, cls).__new__(cls)  # type: ignore
```

test.py
```
from __future__ import print_function

class M(object):
    def __new__(cls, *args, **kwargs):
        print('new =>', cls, args, kwargs)
        if len(args) == 3 and isinstance(args[1], tuple) and args[1][-1].__class__ is cls:
            # subclassing MockObject
            return type(args[0], (M,), args[2], **kwargs)  # type: ignore
        else:
            return super(M, cls).__new__(cls)  # type: ignore

t1 = 'm = M(1, b=2)'
print('exec:', t1); exec(t1)

print('subclassing M')
class F(M):
    def __init__(self, a, b=0):
        self.a = a
        self.b = b

t2 = 'f = F(1, b=2)'
print('exec:', t2); exec(t2)


print('subclassing m')
class X(m):
    def __init__(self, a, b=0):
        self.a = a
        self.b = b

t3 = 'x = X(1, b=2)'
print('exec:', t3); exec(t3)
```

run with py3
```
$ python3 test.py 
exec: m = M(1, b=2)
new => <class '__main__.M'> (1,) {'b': 2}
subclassing M
exec: f = F(1, b=2)
new => <class '__main__.F'> (1,) {'b': 2}
subclassing m
new => <class '__main__.M'> ('X', (<__main__.M object at 0x1013b4400>,), {'__module__': '__main__', '__qualname__': 'X', '__init__': <function X.__init__ at 0x1013a98c8>}) {}
exec: x = X(1, b=2)
new => <class '__main__.X'> (1,) {'b': 2}
```

run with py2
```
$ python2 test.py 
exec: m = M(1, b=2)
new => <class '__main__.M'> (1,) {'b': 2}
subclassing M
exec: f = F(1, b=2)
new => <class '__main__.F'> (1,) {'b': 2}
subclassing m
new => <class '__main__.M'> ('X', (<__main__.M object at 0x102132c50>,), {'__module__': '__main__', '__init__': <function __init__ at 0x102135b18>}) {}
exec: x = X(1, b=2)
new => <class '__main__.X'> (1,) {'b': 2}
```

**Review by shimizukawa — CHANGES_REQUESTED** [review]
I've investigated for py2 error.

File: sphinx/ext/autodoc/importer.py (hunk 2 above)
Comment (tk0miya) [reply sphinx/ext/autodoc/importer.py]: Right. I'd forgotten the usage of `super()` in py2...

File: sphinx/ext/autodoc/importer.py [hunk 3]
```diff
@@ -47,6 +55,10 @@ def __iter__(self):
         # type: () -> None
         pass
 
+    def __mro_entries__(self, bases):
```
Comment (tk0miya) [inline sphinx/ext/autodoc/importer.py:58]: I added this method for https://www.python.org/dev/peps/pep-0560/

**Review by shimizukawa — APPROVED** [review]
LGTM!!

**Comment by tk0miya** [comment]
Thanks!

**Pull request merged**